packages/server-wallet/src/engine/engine.ts

```diff
@@ -551,20 +552,36 @@ export class SingleThreadedEngine
     return response.multipleChannelOutput();
   }
 
+  private async approveObjective(objectiveId: string): Promise<WalletObjective> {
+    return this.store.transaction(async tx => {
+      const objective = await this.store.getAndLockObjective(objectiveId, tx);
+      if (objective.status === 'pending') {
+        const approved = await this.store.approveObjective(objectiveId, tx);
+        await this.registerChannelWithChainService(approved.data.targetChannelId, tx);
+        return approved;
+      } else {
+        return objective;
+      }
+    });
+  }
+
   async approveObjectives(
     objectiveIds: string[]
   ): Promise<{objectives: WalletObjective[]; messages: Message[]}> {
-    const objectives = await this.store.getObjectivesByIds(objectiveIds);
     const channelIds: string[] = [];
     const response = EngineResponse.initialize();
-    for (const objective of objectives) {
-      const {targetChannelId} = objective.data;
-      channelIds.push(targetChannelId);
-      await this.registerChannelWithChainService(targetChannelId);
-
-      await this.store.approveObjective(objective.objectiveId);
+    for (const objectiveId of objectiveIds) {
+      const result = await this.approveObjective(objectiveId);
```

Comment:
> At first, I was concerned that if I call
> `approveObjectives([1,2,3,])` and `approveObjectives([3,2,1])` in quick succession, this would create a deadlock.

Since each objective gets its own transaction I think this avoids the class deadlock scenario when you try to update multiple resources in one transaction. IE: The first call acquires lock on 1, blocks on 3. The second call acquires lock on 3, blocks on 1.

> Should we be worried about deadlocks in this changeset? Can we document why we don't think there are deadlocks?

> It's quite a deviation from the status quo, where we consistently lock on channels. If we need to update the channels table within the `approveObjective` transaction, for instance, we may introduce deadlocks.

That's a fair point! I've updated the code to lock on the `channelId` instead.

 However, I think this would only be an issue if we have two different objectives pointing to the same channelId? If we assume that a "channel has at most one active objective" then wouldn't locking objectiveId be enough?

Interestingly the objective-manager acquires a lock on [objectiveId](https://github.com/statechannels/statechannels/blob/master/packages/server-wallet/src/objectives/objective-manager.ts#L67) and then the objective crankers will lock on the [channelId] (https://github.com/statechannels/statechannels/blob/0563c108f46b3d782a4d0b11f5ffa3e163fa70d8/packages/server-wallet/src/protocols/channel-opener.ts#L46)

Reply:
It's non trivial for me to think through the implications of that 🤣

Reply:
But I think that this pull request does nothing to change the threat of deadlocks.

It seems like we might be relying on the belief that multiple objectives won't be approved for a given channel, but even if we did that, I don't see an obvious threat of deadlocks.